### PR TITLE
fix(synthetics): Add better Error handling for device Emulation scenarios

### PR DESCRIPTION
**Pull request by pjarugula**
# Description

Previously the error isn't appropriate and it makes user difficult to understand what the exact error is, with certain code changes now it is made it into more readable and understandable format.

image
`syntheticsCreateScriptBrowserMonitor` is the NerdGraph mutation we're using from the Terraform resource to create the monitor. if `deviceEmulation` has to be used, `deviceOrientation` and `deviceType` both are compulsory(you can see that `!` mark used at the end of the word)

Fixes # (issue)

Before:
Screenshot 2024-01-18 at 2 29 55 PM

After:
Screenshot 2024-01-18 at 2 31 41 PM

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

Please delete options that are not relevant.

- [ ] My commit message follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] My code is formatted to [Go standards](https://go.dev/blog/gofmt)
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes. Go [here](https://github.com/newrelic/terraform-provider-newrelic/blob/main/CONTRIBUTING.md#testing) for instructions on running tests locally.

## How to test this change?

Please describe how to test your changes. Include any relevant steps in the UI, HCL file(s), commands, etc

- Step 1
- Step 2
- etc

### CODE REVIEW

File: newrelic/structures_newrelic_synthetics_simple_browser_monitor.go [hunk 1]
```diff
@@ -148,17 +155,19 @@ func buildSyntheticsSimpleBrowserMonitorUpdateStruct(d *schema.ResourceData) syn
...
 
-	if doOk || dtOk {
-		simpleBrowserMonitorUpdateInput.AdvancedOptions.DeviceEmulation = &synthetics.SyntheticsDeviceEmulationInput{}
-
-		if doOk {
-			simpleBrowserMonitorUpdateInput.AdvancedOptions.DeviceEmulation.DeviceOrientation = synthetics.SyntheticsDeviceOrientation(do.(string))
-		}
+	if !(scriptLangOk && runtimeTypeOk && runtimeTypeVersionOk) && (doOk && dtOk) {
+		return simpleBrowserMonitorUpdateInput, errors.New("device emulation is not supported by legacy runtime")
+	}
```
Comment (pranav-new-relic) [inline newrelic/structures_newrelic_synthetics_simple_browser_monitor.go:160]: `scriptLangOk` might need to be excluded from this condition.

I find that specifying both `runtime_type` and `runtime_version` without `script_language` in the configuration does not lead to an error (the Synthetics API seems to default an unspecified `scriptLanguage` to `JAVASCRIPT` in the backend) and hence, we don't get a `Bad Runtime Values` error from the Backend like we do if `runtime_type` is specified and `runtime_version` is not; hence the above comment.

File: newrelic/structures_newrelic_synthetics_simple_browser_monitor.go [hunk 2]
```diff
@@ -69,22 +72,26 @@ func buildSyntheticsSimpleBrowserMonitor(d *schema.ResourceData) synthetics.Synt
...
 
-	if doOk || dtOk {
-		simpleBrowserMonitorInput.AdvancedOptions.DeviceEmulation = &synthetics.SyntheticsDeviceEmulationInput{}
-
-		if doOk {
-			simpleBrowserMonitorInput.AdvancedOptions.DeviceEmulation.DeviceOrientation = synthetics.SyntheticsDeviceOrientation(do.(string))
-		}
+	if !(scriptLangOk && runtimeTypeOk && runtimeTypeVersionOk) && (doOk && dtOk) {
+		return simpleBrowserMonitorInput, errors.New("device emulation is not supported by legacy runtime")
+	}
```
Comment (pranav-new-relic) [inline newrelic/structures_newrelic_synthetics_simple_browser_monitor.go:77]: https://github.com/newrelic/terraform-provider-newrelic/pull/2547/files#r1461354914

File: newrelic/structures_newrelic_synthetics_simple_browser_monitor.go [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
 import (
+	"errors"
+
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/newrelic/newrelic-client-go/v2/pkg/synthetics"
 )
 
-func buildSyntheticsSimpleBrowserMonitor(d *schema.ResourceData) synthetics.SyntheticsCreateSimpleBrowserMonitorInput {
+// TODO: Reduce the cyclomatic complexity of this func
+// nolint: gocyclo
```
Comment (pranav-new-relic) [inline newrelic/structures_newrelic_synthetics_simple_browser_monitor.go:11]: I think splitting this function into two can help reduce cyclomatic complexity of this function, for instance, we can shift all of the logic that deals with runtime values and device emulation into a new function. Can you give this a try and see if it works? If not, we can then consider this temporary workaround until we find a better solution

File: newrelic/structures_newrelic_synthetics_simple_browser_monitor.go [hunk 4]
```diff
@@ -69,22 +72,26 @@ func buildSyntheticsSimpleBrowserMonitor(d *schema.ResourceData) synthetics.Synt
...
 		}
 	}
 
-	return simpleBrowserMonitorInput
+	return simpleBrowserMonitorInput, nil
 }
 
-func buildSyntheticsSimpleBrowserMonitorUpdateStruct(d *schema.ResourceData) synthetics.SyntheticsUpdateSimpleBrowserMonitorInput {
+// TODO: Reduce the cyclomatic complexity of this func
+// nolint: gocyclo
```
Comment (pranav-new-relic) [inline newrelic/structures_newrelic_synthetics_simple_browser_monitor.go:93]: I think splitting this function into two can help reduce cyclomatic complexity of this function, for instance, we can shift all of the logic that deals with runtime values and device emulation into a new function. Can you give this a try and see if it works? If not, we can then consider this temporary workaround until we find a better solution

File: newrelic/resource_newrelic_synthetics_monitor_test.go [hunk 5]
```diff
@@ -18,6 +19,97 @@ import (
 
 var tv bool = true
 
+func TestAccNewRelicSyntheticsBrowserMonitor_DeviceEmulationError(t *testing.T) {
```
Comment (pranav-new-relic) [inline newrelic/resource_newrelic_synthetics_monitor_test.go:22]: Can you add a similar test case to test if the error is seen when an existing Terraform configuration is updated?

**Comment by pjarugula** [comment]
Code changes for all the mentioned comments have been done

**Comment by pranav-new-relic** [comment]
Thanks, I'll take a look 👍

**Comment by pranav-new-relic** [comment]
Looks good to me. Nice work 👍

**Review by pranav-new-relic — APPROVED** [review]

**Pull request merged**